public/components/event_analytics/explorer/events_views/data_grid.tsx

```diff
@@ -3,148 +3,271 @@
  */
 
-import React, { useMemo, useState, useEffect, useRef, RefObject } from 'react';
-import { IExplorerFields } from '../../../../../common/types/explorer';
-import { DEFAULT_COLUMNS, PAGE_SIZE } from '../../../../../common/constants/explorer';
-import { getHeaders, getTrs, populateDataGrid } from '../../utils';
+import React, { useMemo, useState, useRef, RefObject, Fragment, useCallback } from 'react';
+import {
+  EuiDataGrid,
+  EuiDescriptionList,
+  EuiDescriptionListDescription,
+  EuiDescriptionListTitle,
+  EuiDataGridColumn,
+} from '@elastic/eui';
+import moment from 'moment';
+import dompurify from 'dompurify';
+import datemath from '@elastic/datemath';
+import { MutableRefObject } from 'react';
+import { GridSortingColumn, IExplorerFields, IField } from '../../../../../common/types/explorer';
+import {
+  DATE_DISPLAY_FORMAT,
+  DATE_PICKER_FORMAT,
+  DEFAULT_SOURCE_COLUMN,
+  DEFAULT_TIMESTAMP_COLUMN,
+} from '../../../../../common/constants/explorer';
 import { HttpSetup } from '../../../../../../../src/core/public';
 import PPLService from '../../../../services/requests/ppl';
+import { FlyoutButton, IDocType } from './docViewRow';
+import { useFetchEvents } from '../../hooks';
+import {
+  PPL_INDEX_INSERT_POINT_REGEX,
+  PPL_NEWLINE_REGEX,
+} from '../../../../../common/constants/shared';
 
 interface DataGridProps {
   http: HttpSetup;
   pplService: PPLService;
-  rows: Array<any>;
-  rowsAll: Array<any>;
+  rows: any[];
+  rowsAll: any[];
   explorerFields: IExplorerFields;
   timeStampField: string;
   rawQuery: string;
+  totalHits: number;
+  requestParams: any;
+  startTime: string;
+  endTime: string;
+  storedSelectedColumns: IField[];
 }
 
 export function DataGrid(props: DataGridProps) {
-  const { http, pplService, rows, rowsAll, explorerFields, timeStampField, rawQuery } = props;
-  const [limit, setLimit] = useState(PAGE_SIZE);
-  const loader = useRef<HTMLDivElement>(null);
-  const [rowRefs, setRowRefs] = useState<RefObject<{closeAllFlyouts(openDocId: string): void}>[]>([]);
-
-  useEffect(() => {
-    if (!loader.current) return;
-
-    const observer = new IntersectionObserver(
-      (entries) => {
-        if (entries[0].isIntersecting) setLimit((limit) => limit + PAGE_SIZE);
-      },
-      {
-        root: null,
-        rootMargin: '500px',
-        threshold: 0,
-      }
-    );
-    observer.observe(loader.current);
-
-    return () => observer.disconnect();
-  }, [loader]);
+  const {
+    http,
+    pplService,
+    rows,
+    rowsAll,
+    explorerFields,
+    timeStampField,
+    rawQuery,
+    totalHits,
+    requestParams,
+    startTime,
+    endTime,
+    storedSelectedColumns,
+  } = props;
+  const [rowRefs, setRowRefs] = useState<
+    Array<RefObject<{ closeAllFlyouts(openDocId: string): void }>>
+  >([]);
+  const { getEvents } = useFetchEvents({
+    pplService,
+    requestParams,
+  });
+  const sortingFields: MutableRefObject<GridSortingColumn[]> = useRef([]);
```

Comment:
why is this a ref? the component should rerender if sort changes

also type would be `MutableRefObject<EuiDataGridSorting['columns']>` if you use the upstream type

Reply:
Changed the type accordingly.
This needs to be a ref because the state won't persist across rerenders otherwise. I tested it out by removing the useRef() (and all associated) portions and an error occurs where the portion on the toolbar the says "Sort fields" doesn't keep track of which fields were sorted (and their directions).

Reply:
There is also an issue with having useState, it manifests itself as sort having a delay. Could be caused by getEvents, data cell rendering, or a few other things. Github Issue will be created after merge.